cache/async_cache.go

```diff
@@ -27,22 +29,45 @@ func (c *AsyncCache) Close() error {
 	return nil
 }
 
-func (c *AsyncCache) AwaitForConcurrentTransaction(key *Key) bool {
-	startTime := time.Now()
+type TransactionResult struct {
+	ElapsedTime time.Duration
+	State       TransactionState
+}
 
+func (c *AsyncCache) AwaitForConcurrentTransaction(key *Key) (TransactionResult, error) {
+	startTime := time.Now()
+	var seenState TransactionState
 	for {
-		if time.Since(startTime) > c.graceTime {
-			// The entry didn't appear during graceTime.
+		elapsedTime := time.Since(startTime)
+		if elapsedTime > c.graceTime {
+			// The entry didn't appear during deadline.
 			// Let the caller creating it.
-			return false
+			return TransactionResult{
+				ElapsedTime: elapsedTime,
+				State:       seenState,
+			}, nil
+		}
+
+		state, err := c.TransactionRegistry.Status(key)
+		if errors.Is(err, ErrMissingTransaction) {
```

Comment:
[opinionated comment]
I don't see the relevance of the error ErrMissingTransaction because:
- you're dealing with it the same as the state transactionCompleted (and when you return the error ErrMissingTransaction you always return an transactionCompleted state)
- it doesn't look like an error since it's the nominal use case

If you want to dissociate the ErrMissingTransaction with the state transactionCompleted you could create a state liketransactionNotPresent. Doing so would allow you to remove this part since it would be handled by the logic of the part if !state.IsPending() {...}

Reply:
Agree with you 👍